### ISSUE TITLE
Neovim blocks in OpenBSD-current (jemalloc infinite loop?)

### ISSUE THREAD

**Issue by equalsraf** [neovim/neovim#3434]
I've tried to build Neovim in OpenBSD (-current Oct-6 snapshot on amd64), the binary is built but running it blocks with no output.

ktrace shows what appears to be an infinite loop that keeps calling `sched_yield` and returning `0`.

From the backtrace this happens at initialization of jemalloc - and disabling jemalloc does fix the issue for me. Sadly I have been unable to isolate the issue further. I've tested with jemalloc 4.0.2 and 4.0.3 and got the same results.

Trace:
```gdb
#0  0x00001efa611309da in sched_yield () at <stdin>:2
#1  0x00001efa7d0caa45 in _spinlock (lock=0x1efa7d4d28d4) at /usr/src/lib/librthread/rthread.c:82
#2  0x00001efa7d0ce06e in _rthread_mutex_lock (mutexp=0x1ef820acb468, trywait=0, abstime=0x0)
    at /usr/src/lib/librthread/rthread_sync.c:98
#3  0x00001ef8205ffc40 in malloc_init_hard () at mutex.h:85
#4  0x00001ef82060726f in calloc (num=1, size=48) at src/jemalloc.c:285
#5  0x00001efa7d0cdf4b in pthread_mutex_init (mutexp=0x1ef820acb468, attr=0x0) at /usr/src/lib/librthread/rthread_sync.c:44
#6  0x00001efa7d0ce083 in _rthread_mutex_lock (mutexp=0x1ef820acb468, trywait=0, abstime=0x0)
    at /usr/src/lib/librthread/rthread_sync.c:100
#7  0x00001ef8205ffc40 in malloc_init_hard () at mutex.h:85
#8  0x00001ef820600cb7 in jemalloc_constructor () at src/jemalloc.c:285
#9  0x00001ef8204340ce in __register_frame_info () from /home/raf/neovim/build/bin/nvim
#10 0x00001ef8204341b6 in atexit () from /home/raf/neovim/build/bin/nvim
#11 0x00001ef820432ea9 in __init () from /home/raf/neovim/build/bin/nvim
#12 0x0000000000000000 in ?? ()
```

The loop is similar to an issue discussed in [openbsd-dev](http://openbsd-archive.7691.n7.nabble.com/process-stuck-in-a-sched-yield-loop-td270329.html) but that issue was already fixed [upstream](http://cvsweb.openbsd.org/cgi-bin/cvsweb/src/lib/librthread/rthread_fork.c.diff?r1=1.11&r2=1.12&f=h). 

The [wiki instructions](https://github.com/neovim/neovim/wiki/Building-Neovim#openbsd--current) for building in OpenBSD say jemalloc does not compile (but it actually does in 4.0.2).

ping @Pyrohh in case you bump into this too.

**Comment by equalsraf** [comment]
The ktrace capture [is here](https://gist.github.com/anonymous/dd2a6c2252ead21f4c50)

**Comment by Pyrohh** [comment]
I've been experiencing this too, so am still using nvim without jemalloc.  IIRC my gdb and ktrace traces were mostly the same as yours, so not much to report there.

This is a bit out of my depth, so I'm unsure whether this should go to tech@, the Jemalloc bug tracker, or if it's an nvim-specific issue.

**Comment by equalsraf** [comment]
> This is a bit out of my depth, so I'm unsure whether this should go to tech@, the Jemalloc bug tracker, or if it's an nvim-specific issue...

Not sure yet, I'm starting to lean towards jemalloc (this seems to start at their [\_\_attribute(constructor)\_\_](https://github.com/jemalloc/jemalloc/blob/21523297fcd72128c14b40ebefbf8ccf114fbede/src/jemalloc.c#L2498)), it can be some conflict with one of our other dependencies.

I have to go now, the best I was able to do was to replicate it by grabbing the Neovim link command and using a test program

```
cc -Wconversion -g -L/usr/local/lib -L/home/raf/Code/neovim/.deps/usr/lib -lintl -liconv -luv -lpthread -lkvm -lkvm -lmsgpack -lluajit-5.1 -lvterm -ltermkey -lunibilium -lpthread -lm -lutil -ljemalloc -Wl,-rpath-link,/usr/X11R6/lib:/usr/local/lib test_jemalloc.c
```

`test_jemalloc.c` is just an empty `main()` function no includes needed. The binary will block.

In fact this works too

```
cc -L/home/raf/Code/neovim/.deps/usr/lib -lpthread -ljemalloc test_jemalloc.c
```

For this test, the problem can be fixed by inverting the order of the linked libraries

```
cc -L/home/raf/Code/neovim/.deps/usr/lib -ljemalloc -lpthread test_jemalloc.c
```

In a nutshell linkage order matters. The same thing also happens with `clang`. I did not get the chance to test a fix with for neovim yet.

**Comment by equalsraf** [comment]
So somewhere between openbsd/jemalloc things go very wrong. The following works

- reorder `-ljemalloc` to be before `-lpthread`
- use `-pthread` instead of `-lpthread` ([more](http://marc.info/?t=123479154600001&r=1&w=2))

**Comment by equalsraf** [comment]
A quick and dirty fix

```diff
diff --git a/src/nvim/CMakeLists.txt b/src/nvim/CMakeLists.txt
index 1038895..453036e 100644
--- a/src/nvim/CMakeLists.txt
+++ b/src/nvim/CMakeLists.txt
@@ -242,7 +242,8 @@ set(NVIM_EXEC_LINK_LIBRARIES ${NVIM_LINK_LIBRARIES})
 
 # Don't use jemalloc in the unit test library.
 if(JEMALLOC_FOUND)
-  list(APPEND NVIM_EXEC_LINK_LIBRARIES ${JEMALLOC_LIBRARIES})
+  # (In OpenBSD) jemalloc must be linked before pthread
+  list(INSERT NVIM 0_EXEC_LINK_LIBRARIES ${JEMALLOC_LIBRARIES})
 endif()
 
 add_executable(nvim ${NEOVIM_GENERATED_SOURCES} ${NEOVIM_SOURCES}
```
Alternatives are:
- use `-pthread` instead of `-lpthread` in OpenBSD (annoying specially because we add `-pthread` multiple times :S)
- Disable jemalloc for OpenBSD

**Comment by Pyrohh** [comment]
That fix above works for me, thanks @equalsraf.

> use -pthread instead of -lpthread in OpenBSD (annoying specially because we add -pthread multiple times :S)

Instead of doing something platform specific, could we just pass `-pthread` to `cc` unconditionally? I'm not sure if our CMake setup causes linking to be done with `cc' or `ld' directly, as the latter doesn't seem to support `-pthread' on OpenBSD.

**Comment by fmoralesc** [comment]
Relevant? https://github.com/jemalloc/jemalloc/commit/d36c7ebb004e73122c76276b854364f543458b8c

**Comment by equalsraf** [comment]
> Instead of doing that, could we just pass -pthread to cc regardless of platform?

Not sure if it is supported everywhere.

> I'm not sure if our CMake setup causes linking to be done with cc or ld directly,

`cc` I believe, at least here.

**Comment by justinmk** [comment]
This was closed by https://github.com/neovim/neovim/pull/3515 and the note at https://github.com/neovim/neovim/wiki/Building-Neovim#openbsd--current , right?

**Comment by justinmk** [comment]
Closed by https://github.com/neovim/neovim/pull/5318
